File: Assets/src/model/enemies/Goo.cs
using UnityEngine;

[System.Serializable]
[ObjectInfo(description: "When attacked, it duplicates into two with half HP.", flavorText: "")]
public class Goo : AIActor, IBodyTakeAttackDamageHandler {
  public override float turnPriority => 50;
  public Goo(Vector2Int pos) : base(pos) {
    faction = Faction.Enemy;
    hp = baseMaxHp = 12;
  }

  internal override (int, int) BaseAttackDamage() => (1, 1);

  protected override ActorTask GetNextTask() {
    var player = GameModel.main.player;
    if (CanTargetPlayer()) {
      if (IsNextTo(player)) {
        return new AttackTask(this, player);
      } else {
        return new ChaseTargetTask(this, player);
      }
    } else {
      return new MoveRandomlyTask(this);
    }
  }

  public void HandleTakeAttackDamage(int damage, int hpBefore, Actor source) {
    if (damage > 0) {
      GameModel.main.EnqueueEvent(Split);
    }
  }

  void Split() {
    if (IsDead) {
      return;
    }
    var hp1 = Mathf.CeilToInt(hp / 2f);
    var hp2 = Mathf.FloorToInt(hp / 2f);

    var goo1 = new Goo(pos);
    goo1.hp = hp1;
    goo1.ClearTasks();
    goo1.statuses.Add(new SurprisedStatus());

    var goo2 = new Goo(pos);
    goo2.hp = hp2;
    goo2.ClearTasks();
    goo2.statuses.Add(new SurprisedStatus());

    floor.PutAll(goo1, goo2);
    // don't die
    floor.Remove(this);
  }
}

File: web/src/model/enemies/Goo.ts
import { AIActor } from './AIActor';
import { ActorTask } from '../ActorTask';
import { AttackTask } from '../tasks/AttackTask';
import { ChaseTargetTask } from '../tasks/ChaseTargetTask';
import { MoveRandomlyTask } from '../tasks/MoveRandomlyTask';
import { SurprisedStatus } from '../tasks/SleepTask';
import { BODY_TAKE_ATTACK_DAMAGE_HANDLER, type IBodyTakeAttackDamageHandler } from '../Body';
import { Vector2Int } from '../../core/Vector2Int';
import { GameModelRef } from '../GameModelRef';
import { entityRegistry } from '../../generator/entityRegistry';
import type { Actor } from '../Actor';

/**
 * When attacked, duplicates into two goos with half HP each.
 * Port of C# Goo.cs.
 */
export class Goo extends AIActor implements IBodyTakeAttackDamageHandler {
  readonly [BODY_TAKE_ATTACK_DAMAGE_HANDLER] = true as const;

  get turnPriority(): number {
    return 50;
  }

  constructor(pos: Vector2Int) {
    super(pos);
    this.hp = this._baseMaxHp = 8;
  }

  baseAttackDamage(): [number, number] {
    return [1, 1];
  }

  handleTakeAttackDamage(damage: number, _hpBefore: number, _source: Actor): void {
    if (damage > 0) {
      GameModelRef.main.enqueuEvent(() => this.split());
    }
  }

  private split(): void {
    if (this.hp < 2) return;

    const hp1 = Math.ceil(this.hp / 2);
    const hp2 = Math.floor(this.hp / 2);

    const goo1 = new Goo(this.pos);
    goo1.hp = hp1;
    goo1.clearTasks();
    goo1.statuses.add(new SurprisedStatus());

    const goo2 = new Goo(this.pos);
    goo2.hp = hp2;
    goo2.clearTasks();
    goo2.statuses.add(new SurprisedStatus());

    this.floor!.putAll([goo1, goo2]);
    this.floor!.remove(this);
  }

  protected getNextTask(): ActorTask {
    const player = GameModelRef.main.player;
    if (this.canTargetPlayer()) {
      if (this.isNextTo(player)) {
        return new AttackTask(this, player);
      }
      return new ChaseTargetTask(this, player);
    }
    return new MoveRandomlyTask(this);
  }
}

entityRegistry.register('Goo', Goo);
